### DISCUSSION TITLE
Fetching version of project in code

### DISCUSSION THREAD

**Discussion by othe1492** [projen/projen#2171] [Q&A]
Hello, we have a CDK construct in Typescript set up with Projen and want to fetch the release version of our construct in the code to use for tagging. Is there any built-in way to do this?

We have tried to replace a placeholder with the version in our pipeline before running the release command but the release will then fail on the git diff since there is a change.

**Comment by mrgrain** [comment]
Hi @othe1492! I'm not sure I understand what exactly you are after. Would you mind clarifying your use case for me?

---

In general, the last released version is available as a tag on the repo. That's were projen is reading it from. The _next_ version is determined by the bump command during the release. It is using `standard-version` in the background and you can configure the bumping implementation using `versionrcOptions` as well as adding [lifecycle scripts](https://github.com/conventional-changelog/standard-version#lifecycle-scripts) to `package.json`.

**Comment by automartin5000** [comment]
I have this same question. I was hoping to set `APP_VERSION` as an env var for use inside the CDK app to tag resources. 

@mrgrain Based on your response, sounds like I'd need to read the git tags myself? However, ideally, this would be code in the Projen project, not in the CDK app. Additionally, as you alluded to, git tags are set _post_-release, I think this might complicate knowing which version to set since it's not immediately obvious to users which version is going to get deployed.

So a couple questions:
1. Is it correct to assume that we need to compute next version ourselves and current version would either need to come from git tags or package.json/pyproject.toml/etc?
2. Is there an easier way for users to understand what the "next" version is going to be?

Sounds like this might also be a good feature request to simplify this?

**Reply by mrgrain** [reply]
Hey @automartin5000 Thanks for adding your use case to this, it helps me understand better what a feature could look like. Although I'm still not exactly sure what it is that both of you are after. I think it might be because you have a specific use-case or release flow in mind?!? Feel free to add any more details you can share.

Anyway, (1) the built-in release system tracks versions exclusively via tags. There is no way to "compute the next version", other than, well, computing it. You should be able to use any versioning tool that can bump versions from tags.

Saying that, in the release workflow a file is created inside a `dist` folder. I guess you could read that. However I gonna say that this is an implementation detail and might change at any time.

**Reply by automartin5000** [reply]
Scenario:
I'm a new user to an existing app. I've checked out the repo which is on version `1.2.0`. 

First thing I maybe want to do is deploy it to my own sandbox account to make sure it's all working. CDK app needs to get the value `1.2.0` without the me having to do anything (git tags I suppose unblocks us, but some helper method/pre-existing env var would be nice here).

Then I want to start working on changes, so I update some code, and want to redeploy, probably from local by directly running `npx projen deploy`. I need the version that gets deployed to change to `1.2.1` or perhaps `1.2.1-alpha1`. I believe the `PRERELEASE` environment variable might help here, but I'm not 100% clear on how. And again, it sounds like in order to get whatever the value is that's set by `PRERELEASE` would need custom code to get computed.

Did I get all that right/Does that help clarify?

**Reply by mrgrain** [reply]
I'm curious, if you are running `npx projen deploy` from your local CDK app (or even a workflow on a feature branch) - what do you need the version for?

**Reply by mrgrain** [reply]
You should be able to use the following workflow:

- `projen bump` (potentially with `PRERELEASE`, which is documented and supported)
- `projen deploy`
- `projen unbump`

**Reply by automartin5000** [reply]
> I'm curious, if you are running `npx projen deploy` from your local CDK app (or even a workflow on a feature branch) - what do you need the version for?

There's probably a few answers here, but for my immediate purposes: debugging with tracing/instrumentation tooling

**Reply by automartin5000** [reply]
> You should be able to use the following workflow:
> 
> * `projen bump` (potentially with `PRERELEASE`, which is documented and supported)
> * `projen deploy`
> * `projen unbump`

Interesting. And then using this flow, the version would end up where? As a tag on the branch? What's the risk if the user forgets to `unbump` before releasing/merging?

**Reply by mrgrain** [reply]
For a TS project `projen bump` sets the version in `package.json`. 

> What's the risk if the user forgets to unbump before releasing/merging?

I'd say pretty low. The default build includes a check if any files change when doing a fresh build. Normally a fresh build will reset the version to `0.0.0`.

**Reply by automartin5000** [reply]
@mrgrain Ok, thanks for the info!
